man/en/sos-upload.1

```diff
@@ -0,0 +1,111 @@
+.TH UPLOAD 1 "July 2024"
+
+.SH NAME
+sos_upload \- Upload files like previously generated sos reports or logs to a policy specific location
+.SH SYNOPSIS
+.B sos upload FILE [options]
+          [--case-id id]\fR
+          [--upload-url url]\fR
+          [--upload-user user]\fR
+          [--upload-pass pass]\fR
+          [--upload-directory dir]\fR
+          [--upload-method]\fR
+          [--upload-no-ssl-verify]\fR
+          [--upload-protocol protocol]\fR
```

Comment:
@jcastill Could the `--upload-protocol s3` flags be included in this work? Unfortunately, it contains unique flags that made S3 easier to implement at the time.
```
  [--upload-s3-endpoint endpoint]
  [--upload-s3-region region]
  [--upload-s3-bucket bucket]
  [--upload-s3-access-key access_key]
  [--upload-s3-secret-key secret_key]
  [--upload-s3-object-prefix object_prefix]
```

----

The existing flags and how the provided values were used were not well aligned for S3, even though valid for FTP, HTTP, etc. protocols. I didn't want to cause any breakage for existing upload protocols while trying to make them work for all protocols, so S3 ended up with unique flags.

I planned to attempt a refactor at some point (sos v5?) where the original protocols and s3 overlap. For example, allowing synonymous flags:
* `--upload-user` ~ `--upload-s3-access-key`
* `--upload-pass` ~ `--upload-s3-secret-key`
* `--upload-directory` ~ `--upload-s3-object-prefix`
* `--upload-url` ~ `--upload-s3-endpoint`

However, I haven't been able to dedicate the time yet.

Reply:
> Could the --upload-protocol s3 flags be included in this work? Unfortunately, it contains unique flags that made S3 easier to implement at the time.

Yes, I'll make sure I include them in the next iteration of this PR.


> I planned to attempt a refactor at some point (sos v5?) where the original protocols and s3 overlap. For example, allowing synonymous flags:
> 
>     --upload-user ~ --upload-s3-access-key
>     --upload-pass ~ --upload-s3-secret-key
>     --upload-directory ~ --upload-s3-object-prefix
>     --upload-url ~ --upload-s3-endpoint
> 
> However, I haven't been able to dedicate the time yet.

Let me know if I can help. My original idea was to have this PR as a starting point and then move stuff out of the generic policy and the OS-specific ones in a second PR, but that was rejected, so I'm working on the full change now. As soon as I finish with that, we can start working on the refactor of S3 it that's OK with you. In fact we need to do some work with S3 uploads for the RH customer portal, so we could do both things in parallel.

Reply:
> Let me know if I can help. My original idea was to have this PR as a starting point and then move stuff out of the generic policy and the OS-specific ones in a second PR, but that was rejected, so I'm working on the full change now. As soon as I finish with that, we can start working on the refactor of S3 it that's OK with you. In fact we need to do some work with S3 uploads for the RH customer portal, so we could do both things in parallel.

When you have a branch published for public view and somewhat functional let me know. I'll branch off of it and start migrating the s3 portions in then submit a PR targeting your branch for you to review.

----

As for the s3 refactoring, we can look into it and I'd be more than happy to try and make some time. I believe a few lend themselves easily, or at least I don't immediately recall any issues with using them, like URL, user, and password. One I do recall bringing up some questions is the `--upload-directory`. For example, should this be only the prefixes inside the bucket? Or should it split the directory like `{bucket}/{prefix}` on only the first slash? There may have been others, but I would have to review the `LinuxPolicy.get_upload_xxxx()` functions and internal `self._vars` again.

Without some "group think" I decided not to implement something I (or others) may have been unhappy with later but stuck with unless making breaking changes.

Reply:
There may be less to refactor than I first thought as I haven't reviewed the code in almost a year. I guess ended up implementing some of it already. Hope I'm still happy with my choices after a year 😄 

https://github.com/sosreport/sos/blob/2aa4fcf0c17f3cf1987ca277069fc09cb0a9f1ae/sos/policies/distros/__init__.py#L615-L629